packages/cli-kit/src/public/node/ui.tsx

```diff
@@ -528,7 +528,65 @@ export async function renderTextPrompt({renderOptions, ...props}: RenderTextProm
 
   // eslint-disable-next-line max-params
   return new Promise((resolve, reject) => {
-    render(<TextPrompt {...props} onSubmit={(value: string) => resolve(value)} />, {
+    render(<TextPrompt {...props} onSubmit={(value: string | undefined) => resolve(value!)} />, {
+      ...renderOptions,
+      exitOnCtrlC: false,
+    })
+      .catch(reject)
+      .finally(resetRecordedSleep)
+  })
+}
+
+export interface RenderDangerousConfirmationPromptOptions extends RenderTextPromptOptions {
+  confirmation: string
+}
+
+/**
+ * Renders a dangerous confirmation prompt to the console, forcing the user to
+ * type a confirmation string to proceed.
+ * @example
+ * ?  Release a new version of nightly-app-2023-06-19?
+ *
+ *    ┃  Includes:
+ *    ┃  + web-px (new)
+ *    ┃  + sub-ui-ext
+ *    ┃  + theme-app-ext
+ *    ┃  + paymentify (from Partner Dashboard)
+ *    ┃
+ *    ┃  Removes:
+ *    ┃  - prod-discount-fun
+ *    ┃
+ *    ┃  This can permanently delete app user data.
+ *
+ *    Type nightly-app-2023-06-19 to confirm, or press Escape
+ *    to cancel.
+ * >
+ *    ▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔▔
+ *
+ */
+export async function renderDangerousConfirmationPrompt({
+  confirmation,
+  renderOptions,
+  ...props
+}: RenderDangerousConfirmationPromptOptions): Promise<boolean> {
+  throwInNonTTY({message: props.message, stdin: renderOptions?.stdin})
+
+  // eslint-disable-next-line prefer-rest-params
+  recordUIEvent({type: 'dangerousConfirmationPrompt', properties: arguments[0]})
+
+  const finalProps = {
+    finalInstruction: ['Type', {userInput: confirmation}, 'to confirm, or press Escape to cancel.'],
+    exitOnEsc: true,
+    validate: (value: string) => {
+      return value === confirmation ? undefined : ['Value must be exactly', {userInput: confirmation}]
```

Comment:
Same consideration here, if there is a passed in `confirmationText` param, couldn't we automatically add this check inside the component itself?

This btw made me think that perhaps we could create a separate component from `TextPrompt` to handle this, something like `DangerousConfirmationPrompt`. The reason is that there is a lot state now in `TextPrompt` that isn't necessary to regular text prompts and I'd expect the return value to be a boolean, but it's still a string instead.

Reply:
A lot of the same logic applies for the actual text prompt piece; the only real difference is validation + `exitOnEsc`. I can try splitting on some reasonable faultline, but if it seems too complicated, I suggest we live with the mess for now and split when it really becomes necessary.

I do like the idea of a text prompt component that returns true/false natively, so it's worth a go.